File: backend/src/agents/executive/engine.ts
import { Act, ActStep, ExecutionReport, ExecutionTask, TaskResult } from '../../schemas/act';
import { randomUUID } from 'node:crypto';

export interface TaskExecutor {
  role: string;
  executeTask(task: ExecutionTask): Promise<TaskResult>;
}

export class ExecutionEngine {
  private _cabinet: Map<string, TaskExecutor>;

  constructor(cabinet: Record<string, TaskExecutor>) {
    this._cabinet = new Map(Object.entries(cabinet));
  }

  public async executeAct(act: Act): Promise<ExecutionReport> {
    const startTime = performance.now();
    const levels = this._topologicalSort(act.steps);
    const results: Map<number, TaskResult> = new Map();
    const failedSteps: Set<number> = new Set();

    for (const level of levels) {
      const toRun: ActStep[] = [];
      
      for (const step of level) {
        // check if dependencies failed
        const blockedDeps = step.dependencies.filter(dep => failedSteps.has(dep));
        if (blockedDeps.length > 0) {
          results.set(step.index, {
            task_id: randomUUID(),
            step_index: step.index,
            status: 'skipped',
            output: `跳过：依赖步骤 [${blockedDeps.join(', ')}] 失败`,
            tokens_consumed: 0,
          });
          failedSteps.add(step.index);
        } else {
          toRun.push(step);
        }
      }

      if (toRun.length > 0) {
        const promises = toRun.map(step => this._executeStep(step, act.act_id));
        const settled = await Promise.allSettled(promises);
        
        for (let i = 0; i < toRun.length; i++) {
          const step = toRun[i];
          const outcome = settled[i];

          if (outcome.status === 'rejected') {
            const errResult: TaskResult = {
              task_id: randomUUID(),
              step_index: step.index,
              status: 'failed',
              error: outcome.reason instanceof Error ? outcome.reason.message : String(outcome.reason),
              output: '',
              tokens_consumed: 0,
            };
            results.set(step.index, errResult);
            failedSteps.add(step.index);
          } else {
            const res = outcome.value;
            results.set(step.index, res);
            if (res.status === 'failed') {
              failedSteps.add(step.index);
            }
          }
        }
      }
    }

    const elapsedMs = performance.now() - startTime;
    const orderedResults: TaskResult[] = [];
    let totalTokens = 0;

    // Ordered output
    const sortedIndices = Array.from(results.keys()).sort((a, b) => a - b);
    for (const idx of sortedIndices) {
      const r = results.get(idx)!;
      orderedResults.push(r);
      if (r.tokens_consumed) {
        totalTokens += r.tokens_consumed;
      }
    }

    const statuses = new Set(orderedResults.map(r => r.status));
    let overall: 'completed' | 'partial' | 'failed';
    if (statuses.size === 1 && statuses.has('success')) {
      overall = 'completed';
    } else if (statuses.has('success')) {
      overall = 'partial';
    } else {
      overall = 'failed';
    }

    return {
      act_id: act.act_id,
      overall_status: overall,
      task_results: orderedResults,
      total_tokens_consumed: totalTokens,
      execution_time_seconds: Number((elapsedMs / 1000).toFixed(3)),
    };
  }

  public resolveSkill(skillName: string): TaskExecutor | undefined {
    return this._cabinet.get(skillName);
  }

  private async _executeStep(step: ActStep, actId: string): Promise<TaskResult> {
    const executor = this.resolveSkill(step.required_skill);
    if (!executor) {
      return {
        task_id: randomUUID(),
        step_index: step.index,
        status: 'failed',
        error: `无法找到 Skill '${step.required_skill}' 对应的执行者`,
        output: '',
        tokens_consumed: 0,
      };
    }

    const task: ExecutionTask = {
      task_id: actId,
      act_id: actId,
      step: step,
      assigned_to: executor.role,
    };

    return await executor.executeTask(task);
  }

  private _topologicalSort(steps: ActStep[]): ActStep[][] {
    const stepMap = new Map<number, ActStep>();
    const inDegree = new Map<number, number>();
    const dependents = new Map<number, number[]>();

    for (const step of steps) {
      stepMap.set(step.index, step);
      inDegree.set(step.index, 0);
      dependents.set(step.index, []);
    }

    for (const step of steps) {
      for (const dep of step.dependencies) {
        if (stepMap.has(dep)) {
          dependents.get(dep)!.push(step.index);
          inDegree.set(step.index, inDegree.get(step.index)! + 1);
        }
      }
    }

    let queue: number[] = [];
    for (const [idx, degree] of inDegree.entries()) {
      if (degree === 0) {
        queue.push(idx);
      }
    }

    const levels: ActStep[][] = [];

    while (queue.length > 0) {
      const currentLevel: ActStep[] = [];
      const nextQueue: number[] = [];

      for (const idx of queue) {
        currentLevel.push(stepMap.get(idx)!);
        for (const depIdx of dependents.get(idx)!) {
          const deg = inDegree.get(depIdx)! - 1;
          inDegree.set(depIdx, deg);
          if (deg === 0) {
            nextQueue.push(depIdx);
          }
        }
      }

      levels.push(currentLevel);
      queue = nextQueue;
    }

    const sortedCount = levels.reduce((acc, level) => acc + level.length, 0);
    if (sortedCount !== steps.length) {
      throw new Error('检测到循环依赖，法案步骤拓扑排序失败');
    }

    return levels;
  }
}

File: openclaw_republic/agents/executive/engine.py
"""行政执行引擎 — 管理法案步骤的顺序/并行执行。

按法案步骤依赖关系进行拓扑排序，同层无依赖的步骤
可并行执行（``asyncio.gather``），跟踪 Token 消耗，
生成 ``ExecutionReport``。
"""

from __future__ import annotations

import asyncio
import time
import uuid
from collections import defaultdict, deque
from typing import Any, Literal, Protocol, runtime_checkable

from openclaw_republic.schemas.act import (
    Act,
    ActStep,
    ExecutionReport,
    ExecutionTask,
    TaskResult,
)


# ---------------------------------------------------------------------------
# 协议：任何拥有 execute_task 的对象均可作为执行者
# ---------------------------------------------------------------------------


@runtime_checkable
class TaskExecutor(Protocol):
    """可执行任务的协议 — 部长 Agent 需满足此接口。"""

    role: str

    async def execute_task(self, task: ExecutionTask) -> TaskResult: ...


# ---------------------------------------------------------------------------
# ExecutionEngine
# ---------------------------------------------------------------------------


class ExecutionEngine:
    """行政执行引擎 — 管理法案步骤的顺序/并行执行。"""

    def __init__(self, cabinet: dict[str, TaskExecutor]) -> None:
        """初始化执行引擎。

        Args:
            cabinet: ``{skill_name: executor}`` 映射，例如
                     ``{"CodeExecution": sec_engineering, "Search": sec_state}``。
        """
        self._cabinet = dict(cabinet)

    # ----- 公开接口 -----

    async def execute_act(self, act: Act, event_publisher: Any | None = None) -> ExecutionReport:
        """按法案步骤列表执行。

        处理逻辑：
        1. 按步骤依赖关系确定执行顺序（拓扑排序）
        2. 同层无依赖的步骤并行执行
        3. 跟踪 Token 消耗
        4. 某步失败时，依赖它的后续步骤标记为 skipped
        5. 生成 ``ExecutionReport``
        """
        start_time = time.monotonic()
        levels = self._topological_sort(act.steps)
        results: dict[int, TaskResult] = {}
        failed_steps: set[int] = set()

        for level in levels:
            # 确定本层哪些步骤需要执行（排除因依赖失败而跳过的）
            to_run: list[ActStep] = []
            for step in level:
                blocked_deps = failed_steps & set(step.dependencies)
                if blocked_deps:
                    # 依赖失败 → 跳过
                    results[step.index] = TaskResult(
                        task_id=str(uuid.uuid4()),
                        step_index=step.index,
                        status="skipped",
                        output=f"跳过：依赖步骤 {sorted(blocked_deps)} 失败",
                    )
                    failed_steps.add(step.index)
                else:
                    to_run.append(step)

            # 并行执行本层待执行步骤
            if to_run:
                coros = [self._execute_step(step, act.act_id, event_publisher) for step in to_run]
                level_results = await asyncio.gather(*coros, return_exceptions=True)
                for step, result in zip(to_run, level_results):
                    if isinstance(result, BaseException):
                        tr = TaskResult(
                            task_id=str(uuid.uuid4()),
                            step_index=step.index,
                            status="failed",
                            error=str(result),
                        )
                        results[step.index] = tr
                        failed_steps.add(step.index)
                    else:
                        results[step.index] = result
                        if result.status == "failed":
                            failed_steps.add(step.index)

        elapsed = time.monotonic() - start_time
        ordered_results = [results[i] for i in sorted(results)]
        total_tokens = sum(r.tokens_consumed for r in ordered_results)

        # 确定整体状态
        statuses = {r.status for r in ordered_results}
        overall: Literal["completed", "partial", "failed"]
        if statuses == {"success"}:
            overall = "completed"
        elif "success" in statuses:
            overall = "partial"
        else:
            overall = "failed"

        return ExecutionReport(
            act_id=act.act_id,
            overall_status=overall,
            task_results=ordered_results,
            total_tokens_consumed=total_tokens,
            execution_time_seconds=round(elapsed, 3),
        )

    def resolve_skill(self, skill_name: str) -> TaskExecutor | None:
        """根据 Skill 名查找对应的内阁部长。"""
        return self._cabinet.get(skill_name)

    # ----- 内部方法 -----

    async def _execute_step(self, step: ActStep, act_id: str, event_publisher: Any | None = None) -> TaskResult:
        """执行单个步骤。"""
        executor = self.resolve_skill(step.required_skill)
        if executor is None:
            return TaskResult(
                task_id=str(uuid.uuid4()),
                step_index=step.index,
                status="failed",
                error=f"无法找到 Skill '{step.required_skill}' 对应的执行者",
            )

        if event_publisher:
            await event_publisher("running", step.required_skill, step.index)

        task = ExecutionTask(
            task_id=str(uuid.uuid4()),
            act_id=act_id,
            step=step,
            assigned_to=executor.role,
        )
        result = await executor.execute_task(task)

        if event_publisher:
            await event_publisher(result.status, step.required_skill, step.index)

        return result

    @staticmethod
    def _topological_sort(steps: list[ActStep]) -> list[list[ActStep]]:
        """将步骤按依赖关系分层（Kahn 拓扑排序）。

        Returns:
            分层结果，每层包含可并行执行的步骤列表。
        """
        step_map: dict[int, ActStep] = {s.index: s for s in steps}
        in_degree: dict[int, int] = {s.index: 0 for s in steps}
        dependents: dict[int, list[int]] = defaultdict(list)

        for s in steps:
            for dep in s.dependencies:
                if dep in step_map:
                    dependents[dep].append(s.index)
                    in_degree[s.index] += 1

        queue: deque[int] = deque(
            idx for idx, deg in in_degree.items() if deg == 0
        )
        levels: list[list[ActStep]] = []

        while queue:
            current_level: list[ActStep] = []
            next_queue: deque[int] = deque()
            while queue:
                idx = queue.popleft()
                current_level.append(step_map[idx])
                for dep_idx in dependents[idx]:
                    in_degree[dep_idx] -= 1
                    if in_degree[dep_idx] == 0:
                        next_queue.append(dep_idx)
            levels.append(current_level)
            queue = next_queue

        return levels
